### ISSUE TITLE
clang 3.6 not supported

### ISSUE THREAD

**Issue by tanner0101** [apple/swift-nio#147]
It seems that clang 3.8 or later is required for use with Swift NIO.

### Environment

```sh
-> /swift/swift-4.1-DEVELOPMENT-SNAPSHOT-2018-03-12-a-ubuntu16.04/usr/bin/swift --version && clang --version
Swift version 4.1-dev (LLVM 1c8b50929b, Clang 420ae40df6, Swift ec821efb01)
Target: x86_64-unknown-linux-gnu
Ubuntu clang version 3.6.2-3ubuntu2 (tags/RELEASE_362/final) (based on LLVM 3.6.2)
Target: x86_64-pc-linux-gnu
Thread model: posix

-> /swift/4.1.0-beta/usr/bin/swift --version && clang --version
Swift version 4.1-dev (LLVM 4d75b446be, Clang db92199adc, Swift 4cb7e7f233)
Target: x86_64-unknown-linux-gnu
Ubuntu clang version 3.6.2-3ubuntu2 (tags/RELEASE_362/final) (based on LLVM 3.6.2)
Target: x86_64-pc-linux-gnu
Thread model: posix
```

```sh
-> lsb_release -a
No LSB modules are available.
Distributor ID:    Ubuntu
Description:    Ubuntu 16.04.2 LTS
Release:    16.04
Codename:    xenial
```

### Error 

```sh
Compile CNIOAtomics src/c-atomics.c
Compile CNIOLinux shim.c
Compile CNIODarwin shim.c
Compile CNIOZlib empty.c
Compile CNIOHTTPParser c_nio_http_parser.c
Compile Swift Module 'Debugging' (4 sources)
In file included from /vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/c-atomics.c:21:
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:20:37: error: expected ';' after top level declarator
struct catmc_atomic__Bool * _Nonnull catmc_atomic__Bool_create(bool value);
                                    ^
                                    ;
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:21:70: error: expected ')'
void catmc_atomic__Bool_destroy(struct catmc_atomic__Bool * _Nonnull atomic);
                                                                     ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:21:32: note: to match this '('
void catmc_atomic__Bool_destroy(struct catmc_atomic__Bool * _Nonnull atomic);
                               ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:22:83: error: expected ')'
bool catmc_atomic__Bool_compare_and_exchange(struct catmc_atomic__Bool * _Nonnull atomic, bool expected, bool desired);
                                                                                  ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:22:45: note: to match this '('
bool catmc_atomic__Bool_compare_and_exchange(struct catmc_atomic__Bool * _Nonnull atomic, bool expected, bool desired);
                                            ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:23:66: error: expected ')'
bool catmc_atomic__Bool_add(struct catmc_atomic__Bool * _Nonnull atomic, bool value);
                                                                 ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:23:28: note: to match this '('
bool catmc_atomic__Bool_add(struct catmc_atomic__Bool * _Nonnull atomic, bool value);
                           ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:24:66: error: expected ')'
bool catmc_atomic__Bool_sub(struct catmc_atomic__Bool * _Nonnull atomic, bool value);
                                                                 ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:24:28: note: to match this '('
bool catmc_atomic__Bool_sub(struct catmc_atomic__Bool * _Nonnull atomic, bool value);
                           ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:25:71: error: expected ')'
bool catmc_atomic__Bool_exchange(struct catmc_atomic__Bool * _Nonnull atomic, bool value);
                                                                      ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:25:33: note: to match this '('
bool catmc_atomic__Bool_exchange(struct catmc_atomic__Bool * _Nonnull atomic, bool value);
                                ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:26:67: error: expected ')'
bool catmc_atomic__Bool_load(struct catmc_atomic__Bool * _Nonnull atomic);
                                                                  ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:26:29: note: to match this '('
bool catmc_atomic__Bool_load(struct catmc_atomic__Bool * _Nonnull atomic);
                            ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:27:68: error: expected ')'
void catmc_atomic__Bool_store(struct catmc_atomic__Bool * _Nonnull atomic, bool value);
                                                                   ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:27:30: note: to match this '('
void catmc_atomic__Bool_store(struct catmc_atomic__Bool * _Nonnull atomic, bool value);
                             ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:29:28: error: redefinition of '_Nonnull' with a different type: 'struct catmc_atomic_char *' vs 'struct catmc_atomic__Bool *'
struct catmc_atomic_char * _Nonnull catmc_atomic_char_create(char value);
                           ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:20:29: note: previous definition is here
struct catmc_atomic__Bool * _Nonnull catmc_atomic__Bool_create(bool value);
                            ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:29:36: error: expected ';' after top level declarator
struct catmc_atomic_char * _Nonnull catmc_atomic_char_create(char value);
                                   ^
                                   ;
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:30:68: error: expected ')'
void catmc_atomic_char_destroy(struct catmc_atomic_char * _Nonnull atomic);
                                                                   ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:30:31: note: to match this '('
void catmc_atomic_char_destroy(struct catmc_atomic_char * _Nonnull atomic);
                              ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:31:81: error: expected ')'
bool catmc_atomic_char_compare_and_exchange(struct catmc_atomic_char * _Nonnull atomic, char expected, char desired);
                                                                                ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:31:44: note: to match this '('
bool catmc_atomic_char_compare_and_exchange(struct catmc_atomic_char * _Nonnull atomic, char expected, char desired);
                                           ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:32:64: error: expected ')'
char catmc_atomic_char_add(struct catmc_atomic_char * _Nonnull atomic, char value);
                                                               ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:32:27: note: to match this '('
char catmc_atomic_char_add(struct catmc_atomic_char * _Nonnull atomic, char value);
                          ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:33:64: error: expected ')'
char catmc_atomic_char_sub(struct catmc_atomic_char * _Nonnull atomic, char value);
                                                               ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:33:27: note: to match this '('
char catmc_atomic_char_sub(struct catmc_atomic_char * _Nonnull atomic, char value);
                          ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:34:69: error: expected ')'
char catmc_atomic_char_exchange(struct catmc_atomic_char * _Nonnull atomic, char value);
                                                                    ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:34:32: note: to match this '('
char catmc_atomic_char_exchange(struct catmc_atomic_char * _Nonnull atomic, char value);
                               ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:35:65: error: expected ')'
char catmc_atomic_char_load(struct catmc_atomic_char * _Nonnull atomic);
                                                                ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:35:28: note: to match this '('
char catmc_atomic_char_load(struct catmc_atomic_char * _Nonnull atomic);
                           ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:36:66: error: expected ')'
void catmc_atomic_char_store(struct catmc_atomic_char * _Nonnull atomic, char value);
                                                                 ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:36:29: note: to match this '('
void catmc_atomic_char_store(struct catmc_atomic_char * _Nonnull atomic, char value);
                            ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:38:29: error: redefinition of '_Nonnull' with a different type: 'struct catmc_atomic_short *' vs 'struct catmc_atomic__Bool *'
struct catmc_atomic_short * _Nonnull catmc_atomic_short_create(short value);
                            ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:20:29: note: previous definition is here
struct catmc_atomic__Bool * _Nonnull catmc_atomic__Bool_create(bool value);
                            ^
/vapor/hi-nodes-production/code/.build/checkouts/swift-nio.git-8876730210459286346/Sources/CNIOAtomics/src/../include/c-atomics.h:38:37: error: expected ';' after top level declarator
struct catmc_atomic_short * _Nonnull catmc_atomic_short_create(short value);
                                    ^
                                    ;
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
error: terminated(1): /swift/4.1.0-beta/usr/bin/swift-build-tool -f /vapor/hi-nodes-production/code/.build/release.yaml main output:


Fetching https://github.com/vapor/console.git
Fetching https://github.com/vapor/database-kit.git
Fetching https://github.com/vapor/routing.git
Fetching https://github.com/apple/swift-nio.git
Fetching https://github.com/vapor/service.git
Fetching https://github.com/vapor/validation.git
Fetching https://github.com/vapor/crypto.git
Fetching https://github.com/apple/swift-nio-zlib-support.git
Fetching https://github.com/vapor/template-kit.git
Fetching https://github.com/vapor/vapor.git
Fetching https://github.com/vapor/core.git
Fetching https://github.com/vapor/engine.git
Fetching https://github.com/vapor/copenssl.git
Cloning https://github.com/apple/swift-nio-zlib-support.git
Resolving https://github.com/apple/swift-nio-zlib-support.git at 1.0.0
Cloning https://github.com/vapor/vapor.git
Resolving https://github.com/vapor/vapor.git at nio
Cloning https://github.com/vapor/core.git
Resolving https://github.com/vapor/core.git at nio
Cloning https://github.com/vapor/copenssl.git
Resolving https://github.com/vapor/copenssl.git at 1.0.0-rc.1
Cloning https://github.com/vapor/console.git
Resolving https://github.com/vapor/console.git at nio
Cloning https://github.com/vapor/database-kit.git
Resolving https://github.com/vapor/database-kit.git at nio
Cloning https://github.com/vapor/routing.git
Resolving https://github.com/vapor/routing.git at nio
Cloning https://github.com/apple/swift-nio.git
Resolving https://github.com/apple/swift-nio.git at 1.1.0
Cloning https://github.com/vapor/service.git
Resolving https://github.com/vapor/service.git at nio
Cloning https://github.com/vapor/validation.git
Resolving https://github.com/vapor/validation.git at nio
Cloning https://github.com/vapor/crypto.git
Resolving https://github.com/vapor/crypto.git at nio
Cloning https://github.com/vapor/template-kit.git
Resolving https://github.com/vapor/template-kit.git at nio
Cloning https://github.com/vapor/engine.git
Resolving https://github.com/vapor/engine.git at nio

Error: deploy failed.
```

**Comment by weissi** [comment]
thanks @tanner0101 I will look into that. I'll likely be tomorrow though until this is resolved.

**Comment by weissi** [comment]
- clang 3.7 already works.
- clang 3.6 has two problems:
  - `_Nonnull` is not defined (which can easily be fixed by `#define _Nonnull __attribute__((nonnull))`
  - clang importer doesn't seem to import nonnull attributes properly so they'll still get IUOs instead of straight types

**Comment by weissi** [comment]
should be fixed in #154

**Comment by Lukasa** [comment]
Resolved by #154.
